core/node/da_clients/src/eigen/verifier.rs

```diff
@@ -0,0 +1,1347 @@
+            .await
+            .map_err(ServiceManagerError::EnrichedClient)?
+            .as_u64();
+
+        let depth = self
+            .cfg
+            .settlement_layer_confirmation_depth
+            .saturating_sub(1);
+        let block_to_return = latest.saturating_sub(depth as u64);
+        Ok(block_to_return)
+    }
+
+    async fn call_batch_id_to_metadata_hash(
+        &self,
+        blob_info: &BlobInfo,
+    ) -> Result<Vec<u8>, VerificationError> {
+        let context_block = self.get_context_block().await?;
+
+        let func_selector =
+            ethabi::short_signature("batchIdToBatchMetadataHash", &[ParamType::Uint(32)]);
+        let mut data = func_selector.to_vec();
+        let mut batch_id_vec = [0u8; 32];
+        U256::from(blob_info.blob_verification_proof.batch_id).to_big_endian(&mut batch_id_vec);
+        data.append(batch_id_vec.to_vec().as_mut());
+
+        let call_request = CallRequest {
+            to: Some(self.cfg.svc_manager_addr),
+            data: Some(zksync_basic_types::web3::Bytes(data)),
+            ..Default::default()
+        };
+
+        let res = self
+            .signing_client
+            .as_ref()
+            .call_contract_function(
+                call_request,
+                Some(BlockId::Number(BlockNumber::Number(context_block.into()))),
+            )
+            .await
+            .map_err(ServiceManagerError::EnrichedClient)?;
+
+        Ok(res.0.to_vec())
+    }
+
+    /// Verifies the certificate batch hash
+    pub(crate) async fn verify_batch(&self, blob_info: &BlobInfo) -> Result<(), VerificationError> {
+        let expected_hash = self.call_batch_id_to_metadata_hash(blob_info).await?;
+
+        if expected_hash == vec![0u8; 32] {
+            return Err(VerificationError::EmptyHash);
+        }
+
+        let actual_hash = self.hash_batch_metadata(
+            &blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .batch_header,
+            &blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .signatory_record_hash,
+            blob_info
+                .blob_verification_proof
+                .batch_medatada
+                .confirmation_block_number,
+        );
+
+        if expected_hash != actual_hash {
+            return Err(VerificationError::DifferentHashes {
+                expected: hex::encode(&expected_hash),
+                actual: hex::encode(&actual_hash),
+            });
+        }
+        Ok(())
+    }
+
+    fn decode_bytes(&self, encoded: Vec<u8>) -> Result<Vec<u8>, VerificationError> {
+        let output_type = [ParamType::Bytes];
+        let tokens: Vec<Token> = ethabi::decode(&output_type, &encoded)
+            .map_err(|e| ServiceManagerError::Decoding(e.to_string()))?;
+        let token = tokens
+            .first()
+            .ok_or(ServiceManagerError::Decoding("No tokens found".to_string()))?;
```

Comment:
Nit: Since `ethabi::decode()` checks tokens against the provided types, it's possible to `unwrap()` here and below instead of returning errors. Also, `tokens.into_iter().next()` would take the first token by value, so it's not necessary to copy bytes below.

Reply:
Fixed here: https://github.com/lambdaclass/zksync-era/pull/386